json-schema-serializer/src/main/java/io/confluent/kafka/formatter/json/JsonSchemaMessageFormatter.java

```diff
@@ -0,0 +1,234 @@
+            || keyDeserializer instanceof AbstractKafkaJsonSchemaDeserializer) {
+          printKeyId = true;
+        }
+      }
+    }
+    if (props.containsKey("schema.id.separator")) {
+      idSeparator = props.getProperty("schema.id.separator").getBytes(StandardCharsets.UTF_8);
+    }
+  }
+
+  private Map<String, Object> getPropertiesMap(Properties props) {
+    Map<String, Object> originals = new HashMap<>();
+    for (final String name : props.stringPropertyNames()) {
+      originals.put(name, props.getProperty(name));
+    }
+    return originals;
+  }
+
+  @Override
+  public void writeTo(ConsumerRecord<byte[], byte[]> consumerRecord, PrintStream output) {
+    if (printTimestamp) {
+      try {
+        TimestampType timestampType = consumerRecord.timestampType();
+        if (timestampType != TimestampType.NO_TIMESTAMP_TYPE) {
+          output.write(String.format("%s:%d", timestampType, consumerRecord.timestamp())
+              .getBytes(StandardCharsets.UTF_8));
+        } else {
+          output.write("NO_TIMESTAMP".getBytes(StandardCharsets.UTF_8));
+        }
+        output.write(keySeparator);
+      } catch (IOException ioe) {
+        throw new SerializationException("Error while formatting the timestamp", ioe);
+      }
+    }
+    if (printKey) {
+      try {
+        if (keyDeserializer != null) {
+          Object deserializedKey = consumerRecord.key() == null
+                                   ? null
+                                   : keyDeserializer.deserialize(null, consumerRecord.key());
+          output.write(deserializedKey != null ? deserializedKey.toString()
+              .getBytes(StandardCharsets.UTF_8) : NULL_BYTES);
+        } else {
+          writeTo(consumerRecord.key(), output);
+        }
+        if (printKeyId) {
+          output.write(idSeparator);
+          int schemaId = schemaIdFor(consumerRecord.key());
+          output.print(schemaId);
+        }
+        output.write(keySeparator);
+      } catch (IOException ioe) {
+        throw new SerializationException("Error while formatting the key", ioe);
+      }
+    }
+    try {
+      writeTo(consumerRecord.value(), output);
+      if (printValueId) {
+        output.write(idSeparator);
+        int schemaId = schemaIdFor(consumerRecord.value());
+        output.print(schemaId);
+      }
+      output.write(lineSeparator);
+    } catch (IOException ioe) {
+      throw new SerializationException("Error while formatting the value", ioe);
+    }
+  }
+
+  private void writeTo(byte[] data, PrintStream output) throws IOException {
+    Object object = deserialize(data);
+    output.print(objectMapper.writeValueAsString(object));
+  }
+
+  @Override
+  public void close() {
+    // nothing to do
+  }
+
+  private int schemaIdFor(byte[] payload) {
+    ByteBuffer buffer = ByteBuffer.wrap(payload);
+    if (buffer.get() != MAGIC_BYTE) {
+      throw new SerializationException("Unknown magic byte!");
+    }
+    return buffer.getInt();
+  }
```

Comment:
I feel like this could be refactored out to a utility class, if it doesn't already exist